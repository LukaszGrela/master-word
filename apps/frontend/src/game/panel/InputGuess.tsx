import { FC, useCallback, useEffect, useState } from 'react';
import { createPortal } from 'react-dom';
import { isCorrectWord, useRunAfterUpdate, noop } from '@repo/utils';
import { TErrorResponse } from '../../api';
import { EditIcon } from '../icons/EditIcon';
import { useLanguage } from '../../i18n';
import './InputGuess.css';

type TOnCloseGuess = (action: 'guess', word: string) => void;
type TOnCloseStart = (action: 'start' | 'close') => void;
type TOnClose = TOnCloseGuess & TOnCloseStart;

const InputGuess: FC<{
  show: boolean;

  initWord: string;
  maxLength: number;
  onClose: TOnClose;
  error?: (TErrorResponse & { details?: string }) | null;
}> = ({ maxLength, initWord = '', onClose, error, show }) => {
  const { getUIText: t } = useLanguage();
  const [guess, setGuess] = useState(initWord);
  const [displayed, setDisplayed] = useState(show);
  const [hidden, setHidden] = useState(false);
  useEffect(() => {
    setDisplayed(show);
    if (show) {
      setGuess(initWord);
    } else {
      setHidden(false);
      setGuess('');
    }
  }, [show, initWord]);

  useEffect(() => {
    if (error?.code === 6 && error?.details) {
      setGuess(error.details);
    }
  }, [error?.code, error?.details]);

  const handleCommit = useCallback(() => {
    onClose('guess', guess);
  }, [guess, onClose]);

  const runAfterUpdate = useRunAfterUpdate();
  const handleChangeText: React.ChangeEventHandler<HTMLInputElement> =
    useCallback(
      (e) => {
        const input = e.target;
        const value = input.value;
        const { selectionEnd } = input;
        if (value === '' || isCorrectWord(value)) {
          const newValue = value.toLocaleUpperCase();
          let setCursor = noop;

          if (guess !== newValue && value !== '') {
            setCursor = () => {
              input.selectionStart = selectionEnd;
              input.selectionEnd = selectionEnd;
            };
          }
          setGuess(newValue);
          runAfterUpdate(setCursor);
        }
      },
      [guess, runAfterUpdate],
    );

  const disableButton = guess.length < maxLength;
  const handleEnter: React.KeyboardEventHandler<HTMLInputElement> = useCallback(
    (e) => {
      if (e.key === 'Enter' && !disableButton) {
        // commit
        onClose('guess', guess);
      }
    },
    [disableButton, guess, onClose],
  );

  const handleClose = useCallback(() => {
    setHidden(true);
  }, []);
  const handleShow = useCallback(() => {
    setHidden(false);
  }, []);

  useEffect(() => {
    const tapHandler = (e: MouseEvent) => {
      if ((e.target as HTMLElement).classList.contains('board')) {
        setHidden(false);
      }
    };

    if (show && hidden) {
      document.addEventListener('click', tapHandler);
    }

    return () => {
      document.removeEventListener('click', tapHandler);
    };
  }, [hidden, show]);

  return (
    displayed &&
    (hidden ? (
      createPortal(
        <div className="input-modal fab-container">
          <button
            data-testid="input-modal-fab-button"
            className="fab"
            onClick={handleShow}
          >
            <EditIcon />
          </button>
        </div>,
        document.body,
      )
    ) : (
      <div className="panel input-modal">
        <label htmlFor="guess">{t('input-modal-label')}</label>
        <input
          data-testid="input-modal-guess"
          type="text"
          name="guess"
          maxLength={maxLength}
          autoFocus
          value={guess}
          onChange={handleChangeText}
          /* v8 ignore next */
          onPaste={(e) => e.preventDefault()}
          onKeyUp={handleEnter}
          autoComplete="off"
          enterKeyHint="done"
          aria-errormessage={error?.code === 6 ? error?.error : undefined}
          aria-invalid={error?.code === 6}
        ></input>
        {!disableButton && (
          <div className="read-the-docs">{t('input-modal-info-done')}</div>
        )}
        {disableButton && (
          <div className="read-the-docs">
            {t('input-modal-info-start', { maxLength })}
          </div>
        )}
        <div className="button-row">
          <button className="secondary" onClick={handleClose}>
            {t('input-modal-close-button')}
          </button>
          <button
            className="primary"
            onClick={handleCommit}
            disabled={disableButton}
          >
            {t('input-modal-ok-button')}
          </button>
        </div>
      </div>
    ))
  );
};

export default InputGuess;
