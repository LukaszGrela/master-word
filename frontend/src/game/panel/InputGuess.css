.input-modal input[type='text'] {
  border-radius: 8px;
  border: 1px solid transparent;

  font-size: 1rem;
  font-weight: 500;
  font-family: inherit;
  background-color: #1a1a1a;

  transition: border-color 0.25s;

  padding: 0.5rem;
  margin-bottom: 0.5rem;

  max-width: 12rem;
}
.input-modal input[type='text']:hover {
  border-color: #646cff;
}
.input-modal input[type='text']:focus,
.input-modal input[type='text']:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.input-modal input[type='text']:invalid {
  border-color: red;
}

.input-modal label {
  font-weight: 500;
  font-size: large;
  margin: 0.5rem 0;
  display: block;
}

@media (prefers-color-scheme: light) {
  .input-modal input[type='text'] {
    background-color: #f9f9f9;
  }
}
