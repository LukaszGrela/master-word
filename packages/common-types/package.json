{
  "name": "@repo/common-types",
  "version": "0.0.2",
  "private": true,
  "license": "MIT",
  "exports": {
    ".": "./src/index.ts"
  },
  "devDependencies": {
    "@repo/config-typescript": "*",
    "eslint": "^8.56.0",
    "typescript": "^5.2.2"
  }
}
