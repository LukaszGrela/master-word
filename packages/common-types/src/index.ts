/**
 *
 */
export interface IDictionary<T = unknown> {
  [key: string]: T;
}

/**
 * https://greladesign.co/blog/2024/02/19/typescript-make-stricter-required
 */
export type TStrictRequired<T, E = null | undefined> = {
  [P in keyof T]-?: Exclude<T[P], E>;
};
export type PartialPick<T, K extends keyof T> = Partial<Omit<T, K>> &
  Pick<T, K>;
