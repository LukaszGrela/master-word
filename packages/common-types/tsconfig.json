{
  "extends": "@repo/config-typescript/tsconfig.base.json",
  "include": [
    "src/**/*"
  ],
  "exclude": [
    "nod_modules",
    "src/test/**/*",
    "src/**/*.spec.ts"
  ]
}
