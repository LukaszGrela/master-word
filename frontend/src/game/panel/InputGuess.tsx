import { FC, useCallback, useEffect, useState } from 'react';
import t from '../../i18n';
import { TErrorResponse } from '../../api';
import { isCorrectWord } from '../../utils/isLetter';
import { EditIcon } from '../icons/EditIcon';
import { createPortal } from 'react-dom';
import './InputGuess.css';

type TOnCloseGuess = (action: 'guess', word: string) => void;
type TOnCloseStart = (action: 'start' | 'close') => void;
type TOnClose = TOnCloseGuess & TOnCloseStart;

const InputGuess: FC<{
  show: boolean;

  initWord: string;
  maxLength: number;
  onClose: TOnClose;
  error?: (TErrorResponse & { details?: string }) | null;
}> = ({ maxLength, initWord = '', onClose, error, show }) => {
  const [guess, setGuess] = useState(initWord);
  const [displayed, setDisplayed] = useState(show);
  const [hidden, setHidden] = useState(false);
  useEffect(() => {
    setDisplayed(show);
    if (show) {
      setGuess(initWord);
    } else {
      setHidden(false);
      setGuess('');
    }
  }, [show, initWord]);

  useEffect(() => {
    if (error?.code === 6 && error?.details) {
      setGuess(error.details);
    }
  }, [error?.code, error?.details]);

  const handleCommit = useCallback(() => {
    onClose('guess', guess);
  }, [guess, onClose]);

  const handleChangeText: React.ChangeEventHandler<HTMLInputElement> =
    useCallback((e) => {
      const value = e.target.value;
      if (value === '' || isCorrectWord(value)) {
        setGuess(value.toLocaleUpperCase());
      }
    }, []);

  const disableButton = guess.length < maxLength;
  const handleEnter: React.KeyboardEventHandler<HTMLInputElement> = useCallback(
    (e) => {
      if (e.key === 'Enter' && !disableButton) {
        // commit
        onClose('guess', guess);
      }
    },
    [disableButton, guess, onClose]
  );

  const handleClose = useCallback(() => {
    setHidden(true);
  }, []);
  const handleShow = useCallback(() => {
    setHidden(false);
  }, []);

  useEffect(() => {
    const tapHandler = (e: MouseEvent) => {
      if ((e.target as HTMLElement).classList.contains('board')) {
        setHidden(false);
      }
    };

    if (show && hidden) {
      document.addEventListener('click', tapHandler);
    }

    return () => {
      document.removeEventListener('click', tapHandler);
    };
  }, [hidden, show]);

  return (
    displayed &&
    (hidden ? (
      createPortal(
        <div className='input-modal fab-container'>
          <button className='fab' onClick={handleShow}>
            <EditIcon />
          </button>
        </div>,
        document.body
      )
    ) : (
      <div className='panel input-modal'>
        <label htmlFor='guess'>{t('input-modal-label')}</label>
        <input
          type='text'
          name='guess'
          maxLength={maxLength}
          autoFocus
          value={guess}
          onChange={handleChangeText}
          onPaste={(e) => e.preventDefault()}
          onKeyUp={handleEnter}
          autoComplete='off'
          enterKeyHint='done'
          aria-errormessage={error?.code === 6 ? error?.error : undefined}
          aria-invalid={error?.code === 6}
        ></input>
        {!disableButton && (
          <div className='read-the-docs'>{t('input-modal-info-done')}</div>
        )}
        {disableButton && (
          <div className='read-the-docs'>
            {t('input-modal-info-start', { maxLength })}
          </div>
        )}
        <div className='button-row'>
          <button onClick={handleClose}>{t('input-modal-close-button')}</button>
          <button onClick={handleCommit} disabled={disableButton}>
            {t('input-modal-ok-button')}
          </button>
        </div>
      </div>
    ))
  );
};

export default InputGuess;
