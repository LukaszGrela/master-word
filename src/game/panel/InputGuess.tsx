import { FC, useCallback, useState } from 'react';
import './InputGuess.css';

// const reg = /^[\p{Letter}\p{Mark}]+$/iu;

const InputGuess: FC<{
  initWord: string;
  maxLength: number;
  onClose: (action: 'guess' | 'start', word: string) => void;
}> = ({ maxLength, initWord = '', onClose }) => {
  const [guess, setGuess] = useState(initWord);
  const handleCommit = useCallback(() => {
    onClose('guess', guess);
  }, [guess, onClose]);

  const handleChangeText: React.ChangeEventHandler<HTMLInputElement> =
    useCallback((e) => {
      const value = e.target.value;

      setGuess(value.toLocaleUpperCase());
    }, []);

  const disableButton = guess.length < maxLength;
  const handleEnter: React.KeyboardEventHandler<HTMLInputElement> = useCallback(
    (e) => {
      if (e.code === 'Enter' && !disableButton) {
        // commit
        onClose('guess', guess);
      }
    },
    [disableButton, guess, onClose]
  );

  return (
    <div className='panel input-modal'>
      <label htmlFor='guess'>Enter word to guess</label>
      <input
        type='text'
        name='guess'
        maxLength={maxLength}
        autoFocus
        value={guess}
        onChange={handleChangeText}
        onPaste={(e) => e.preventDefault()}
        onKeyUp={handleEnter}
        autoComplete='off'
      ></input>
      {!disableButton && (
        <div className='read-the-docs'>Tap Confirm or Enter to commit</div>
      )}
      {disableButton && (
        <div className='read-the-docs'>Enter 5 letter word to guess</div>
      )}
      <button onClick={handleCommit} disabled={disableButton}>
        Confirm
      </button>
    </div>
  );
};

export default InputGuess;
