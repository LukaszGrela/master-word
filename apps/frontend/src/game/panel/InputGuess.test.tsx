import { describe, expect, it, vi } from 'vitest';
import { fireEvent } from '@testing-library/react';
import {
  contextRenderer,
  defaultLanguageContext,
  MockContext,
} from '../../i18n/LanguageContext/__tests__/contextRenderer';

import InputGuess from './InputGuess';
import { IDictionary } from '@repo/common-types';

describe('game', () => {
  describe('panel', () => {
    describe('InputGuess', () => {
      it('Renders with required props', () => {
        const { container } = contextRenderer(
          <InputGuess
            show
            initWord=""
            maxLength={5}
            onClose={vi.fn<
              [action: 'start' | 'close'] | [action: 'guess', word: string],
              void
            >()}
          />,
        );

        expect(container).toMatchSnapshot();
      });
      it('Renders with context and all props', () => {
        const { container } = contextRenderer(
          <InputGuess
            show
            initWord=""
            maxLength={5}
            onClose={vi.fn<
              [action: 'start' | 'close'] | [action: 'guess', word: string],
              void
            >()}
          />,
          {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: vi.fn().mockImplementation((key: string) => {
                const en: IDictionary<string> = {
                  'input-modal-label': 'Insert word to guess',
                  'input-modal-info-done': 'Tap Done to check',
                  'input-modal-info-start': 'Enter {maxLength} letter word',
                  'input-modal-ok-button': 'Done',
                  'input-modal-close-button': 'Hide',
                };
                return en[key] || key;
              }),
            },
            renderOptions: {},
          },
        );

        expect(container).toMatchSnapshot();
      });
      it('Hide/Show panel', () => {
        const dataTestId = {
          'data-testid': 'board',
        };
        const { queryByText, getByText, queryByTestId, getByTestId } =
          contextRenderer(
            <div>
              <div className="board" {...dataTestId}>
                Click Me
              </div>
              <InputGuess
                show
                initWord=""
                maxLength={5}
                onClose={vi.fn<
                  [action: 'start' | 'close'] | [action: 'guess', word: string],
                  void
                >()}
              />
            </div>,
            {
              languageContext: {
                ...defaultLanguageContext,
                getUIText: vi.fn().mockImplementation((key: string) => {
                  const en: IDictionary<string> = {
                    'input-modal-label': 'Insert word to guess',
                    'input-modal-info-done': 'Tap Done to check',
                    'input-modal-info-start': 'Enter {maxLength} letter word',
                    'input-modal-ok-button': 'Done',
                    'input-modal-close-button': 'Hide',
                  };
                  return en[key] || key;
                }),
              },
              renderOptions: {},
            },
          );

        let labelBefore = queryByText('Insert word to guess');
        expect(labelBefore).toBeDefined();

        let closeBtn = getByText('Hide');
        fireEvent.click(closeBtn);

        const label = queryByText('Insert word to guess');
        expect(label).toBeNull();

        // FAB should exist
        let fab = queryByTestId('input-modal-fab-button');
        expect(fab).toBeDefined();
        fireEvent.click(fab!);

        labelBefore = queryByText('Insert word to guess');
        expect(labelBefore).toBeDefined();

        // do it again
        closeBtn = getByText('Hide');
        fireEvent.click(closeBtn);
        fab = queryByTestId('input-modal-fab-button');
        expect(fab).toBeDefined();

        // but this time use tap
        const board = getByTestId('board');
        expect(board).toBeDefined();

        fireEvent.click(board);

        // should be opened
        labelBefore = queryByText('Insert word to guess');
        expect(labelBefore).toBeDefined();
      });
      it('Inserts guess', () => {
        const onClose = vi.fn<
          [action: 'start' | 'close'] | [action: 'guess', word: string],
          void
        >();

        const languageContext = {
          ...defaultLanguageContext,
          getUIText: vi.fn().mockImplementation((key: string) => {
            const en: IDictionary<string> = {
              'input-modal-label': 'Insert word to guess',
              'input-modal-info-done': 'Tap Done to check',
              'input-modal-info-start': 'Enter {maxLength} letter word',
              'input-modal-ok-button': 'Done',
              'input-modal-close-button': 'Hide',
            };
            return en[key] || key;
          }),
        };

        const { rerender, queryByText, getByText, getByTestId } =
          contextRenderer(
            <InputGuess show initWord="" maxLength={5} onClose={onClose} />,
            {
              languageContext,
              renderOptions: {},
            },
          );

        let readTheDocs = queryByText('Enter {maxLength} letter word');
        expect(readTheDocs).toBeDefined();

        let closeBtn = getByText('Done');
        expect(closeBtn.getAttribute('disabled')).toBe('');

        const input = getByTestId('input-modal-guess');
        // type word
        fireEvent.change(input, { target: { value: 'GUESS' } });

        // it should now be enabled
        closeBtn = getByText('Done');
        expect(closeBtn.getAttribute('disabled')).toBe(null);

        // read the docs adjusted
        readTheDocs = queryByText('Tap Done to check');
        expect(readTheDocs).toBeDefined();

        // keyboard event
        input.focus();
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 'Enter',
          code: 'Enter',
          charCode: 13,
        });

        // callback called
        expect(onClose).toBeCalledTimes(1);
        expect(onClose).toHaveBeenLastCalledWith('guess', 'GUESS');

        // close modal
        fireEvent.click(closeBtn);

        // callback called
        expect(onClose).toBeCalledTimes(2);
        expect(onClose).toHaveBeenLastCalledWith('guess', 'GUESS');

        // close modal
        rerender(
          <MockContext context={languageContext}>
            <InputGuess
              show={false}
              initWord=""
              maxLength={5}
              onClose={onClose}
            />
          </MockContext>,
        );

        // expect not to find any panel traces
        expect(queryByText('Done')).toBeNull();
      });
      it('Display error', () => {
        const onClose = vi.fn<
          [action: 'start' | 'close'] | [action: 'guess', word: string],
          void
        >();

        const { container } = contextRenderer(
          <InputGuess
            show
            initWord=""
            maxLength={5}
            onClose={onClose}
            error={{
              code: 6,
              details: 'WRONG',
              error: 'Invalid word',
            }}
          />,
          {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: vi.fn().mockImplementation((key: string) => {
                const en: IDictionary<string> = {
                  'input-modal-label': 'Insert word to guess',
                  'input-modal-info-done': 'Tap Done to check',
                  'input-modal-info-start': 'Enter {maxLength} letter word',
                  'input-modal-ok-button': 'Done',
                  'input-modal-close-button': 'Hide',
                };
                return en[key] || key;
              }),
            },
            renderOptions: {},
          },
        );

        expect(container).toMatchSnapshot();
      });
    });
  });
});
