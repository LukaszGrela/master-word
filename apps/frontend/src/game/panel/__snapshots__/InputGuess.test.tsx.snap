// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`game > panel > InputGuess > Display error 1`] = `
<div>
  <div
    class="panel input-modal"
  >
    <label
      for="guess"
    >
      Insert word to guess
    </label>
    <input
      aria-errormessage="Invalid word"
      aria-invalid="true"
      autocomplete="off"
      data-testid="input-modal-guess"
      enterkeyhint="done"
      maxlength="5"
      name="guess"
      type="text"
      value="WRONG"
    />
    <div
      class="read-the-docs"
    >
      Tap Done to check
    </div>
    <div
      class="button-row"
    >
      <button
        class="secondary"
      >
        Hide
      </button>
      <button
        class="primary"
      >
        Done
      </button>
    </div>
  </div>
</div>
`;

exports[`game > panel > InputGuess > Renders with context and all props 1`] = `
<div>
  <div
    class="panel input-modal"
  >
    <label
      for="guess"
    >
      Insert word to guess
    </label>
    <input
      aria-invalid="false"
      autocomplete="off"
      data-testid="input-modal-guess"
      enterkeyhint="done"
      maxlength="5"
      name="guess"
      type="text"
      value=""
    />
    <div
      class="read-the-docs"
    >
      Enter {maxLength} letter word
    </div>
    <div
      class="button-row"
    >
      <button
        class="secondary"
      >
        Hide
      </button>
      <button
        class="primary"
        disabled=""
      >
        Done
      </button>
    </div>
  </div>
</div>
`;

exports[`game > panel > InputGuess > Renders with required props 1`] = `
<div>
  <div
    class="panel input-modal"
  >
    <label
      for="guess"
    />
    <input
      aria-invalid="false"
      autocomplete="off"
      data-testid="input-modal-guess"
      enterkeyhint="done"
      maxlength="5"
      name="guess"
      type="text"
      value=""
    />
    <div
      class="read-the-docs"
    />
    <div
      class="button-row"
    >
      <button
        class="secondary"
      />
      <button
        class="primary"
        disabled=""
      />
    </div>
  </div>
</div>
`;
